packages/next/vitals/vitals.ts

```diff
@@ -0,0 +1,24 @@
+import { useEffect } from 'react'
+import { NextWebVitalsMetric } from '../pages/_app'
+
+type ReportWebVitalsCallback = (webVitals: NextWebVitalsMetric) => any
+export const webVitalsCallbacks = new Set<ReportWebVitalsCallback>()
+const metrics: NextWebVitalsMetric[] = []
+
+export function trackWebVitalMetric(metric: NextWebVitalsMetric) {
+  metrics.push(metric)
+  webVitalsCallbacks.forEach((callback) => callback(metric))
+}
+
+export function useWebVitalsReport(callback: ReportWebVitalsCallback) {
+  useEffect(() => {
+    if (metrics.length) {
+      metrics.forEach((metric) => callback(metric))
+    }
+
+    webVitalsCallbacks.add(callback)
+    return () => {
+      webVitalsCallbacks.delete(callback)
+    }
+  }, [callback])
```

Comment:
My only concern here is using callback as the dependency might cause the effect being triggered on every render, if an inline anonymous function is passed. In that case all the metrics will always get re-reported again.

I think a better way is put the callback in a ref, and get rid of the effect dependency:

```suggestion
  const callbackRef = useRef(callback)
  useEffect(() => callbackRef.current = callback)

  useEffect(() => {
    const handler = metric => callbackRef.current(metric)
   
    metrics.forEach((metric) => handler(metric))

    webVitalsCallbacks.add(handler)
    return () => {
      webVitalsCallbacks.delete(handler)
    }
  }, [])
```